src/com/ms/silverking/cloud/dht/daemon/storage/NamespaceStore.java

```diff
@@ -2683,892 +2893,1199 @@ private void moveToNext() {
-            int    maxValues;
-            int    sIndex;
-            
-            maxValues = numValues(pendingPuts, numPendingPuts);
-            sIndex = 0;
-            while (sIndex < numPendingPuts) {
-                KeyedOpResultMultiplexor    resultListenerMultiplexor;
-                List<StorageValueAndParameters>    values;
-                NamespaceStore    nsStore;
-                
-                nsStore = pendingPuts[sIndex].nsStore;
-                values = new ArrayList<>(maxValues);
-                resultListenerMultiplexor = new KeyedOpResultMultiplexor();
-                while (sIndex < numPendingPuts && pendingPuts[sIndex].nsStore == nsStore) {
-                    for (StorageValueAndParameters svp : pendingPuts[sIndex].values) {
-                        values.add(svp);
-                        resultListenerMultiplexor.addListener(svp.getKey(), pendingPuts[sIndex].resultListener);
-                    }
-                    sIndex++;
-                }
-                if (values.size() <= 0) {
-                    throw new RuntimeException("Unexpected values.size() <= 0: "+ values.size());
-                }
-                if (nsStore.putTrigger != null && nsStore.putTrigger.supportsMerge()) {
-                    Set<Waiter>                triggeredWaitFors;
-                    Map<DHTKey,OpResult>    mergeResults;
-                    
-                    mergeResults = nsStore.putTrigger.mergePut(values);
-                    nsStore.writeLock.lock();
-                    try {
-                        triggeredWaitFors = nsStore.notifyAndCheckWaiters(mergeResults, resultListenerMultiplexor);
-                    } finally {
-                        nsStore.writeLock.unlock();
-                    }
-                    nsStore.handleTriggeredWaitFors(triggeredWaitFors);
-                    
-                    for (Map.Entry<DHTKey, OpResult> resultEntry : mergeResults.entrySet()) {
-                        resultListenerMultiplexor.sendResult(resultEntry.getKey(), resultEntry.getValue());
-                    }
-                } else {
-                    nsStore.put(values, null, resultListenerMultiplexor); // Note: pending puts don't support userdata
-                }
-                maxValues -= values.size();
-            }
+    private void process() {
+      int numTaken;
+
+      try {
+        //numTaken = q.takeMultiple(work);
+        numTaken = takeMultiple();
+        if (Log.levelMet(Level.INFO)) {
+          Log.warningf("q.size() %d\tnumTaken %d", q.size(), numTaken);
         }
-                
-        public void run() {
-            while (true) {
-                try {
-                    process();
-                } catch (Exception e) {
-                    Log.logErrorWarning(e);
-                    ThreadUtil.pauseAfterException();
-                }
-            }
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)